diff --git a/device-common.mk b/device-common.mk
index e092764..e538087 100644
--- a/device-common.mk
+++ b/device-common.mk
@@ -157,6 +157,9 @@ PRODUCT_COPY_FILES +=  \
 # Include BT modules
 $(call inherit-product-if-exists, device/linaro/hikey/wpan/ti-wpan-products.mk)
 
+# Include OP-TEE packages
+$(call inherit-product-if-exists, device/linaro/hikey/optee-packages.mk)
+
 PRODUCT_COPY_FILES += \
         frameworks/native/data/etc/android.hardware.wifi.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.hardware.wifi.xml \
         frameworks/native/data/etc/android.hardware.wifi.direct.xml:$(TARGET_COPY_OUT_VENDOR)/etc/permissions/android.hardware.wifi.direct.xml \
diff --git a/init.common.rc b/init.common.rc
index c15b29d..ddc9800 100644
--- a/init.common.rc
+++ b/init.common.rc
@@ -126,3 +126,14 @@ service bugreport /system/bin/dumpstate -d -p -B -z \
     class main
     disabled
     oneshot
+
+on post-fs-data
+    # create for OP-TEE test
+    mkdir /data/tee 0777 shell shell
+    mkdir /data/tee/optee_armtz 0777 shell shell
+    restorecon_recursive /data/tee
+
+service tee-supplicant /vendor/bin/tee-supplicant
+    class main
+    user root
+    oneshot
diff --git a/optee-packages.mk b/optee-packages.mk
new file mode 100644
index 0000000..3831c0e
--- /dev/null
+++ b/optee-packages.mk
@@ -0,0 +1,46 @@
+OPTEE_PLATFORM ?= hikey
+OPTEE_PLATFORM_FLAVOR ?= hikey960
+OPTEE_CFG_ARM64_CORE ?= y
+OPTEE_TA_TARGETS ?= ta_arm64
+OPTEE_OS_DIR ?= optee/optee_os
+BUILD_OPTEE_MK := $(OPTEE_OS_DIR)/mk/aosp_optee.mk
+
+# OP TEE client library and service
+PRODUCT_PACKAGES += libteec \
+                    tee-supplicant \
+
+# optee_test and TA
+PRODUCT_PACKAGES += xtest
+# os_test_lib
+PRODUCT_PACKAGES += ffd2bded-ab7d-4988-95ee-e4962fff7154.ta
+# os_test
+PRODUCT_PACKAGES += 5b9e0e40-2636-11e1-ad9e-0002a5d5c51b.ta
+# concurrent_large
+PRODUCT_PACKAGES += 5ce0c432-0ab0-40e5-a056-782ca0e6aba2.ta
+# sha_perf
+PRODUCT_PACKAGES += 614789f2-39c0-4ebf-b235-92b32ac107ed.ta
+# storage2
+PRODUCT_PACKAGES += 731e279e-aafb-4575-a771-38caa6f0cca6.ta
+# storage
+PRODUCT_PACKAGES += b689f2a7-8adf-477a-9f99-32e90c0ad0a2.ta
+# create_fail_test
+PRODUCT_PACKAGES += c3f6e2c0-3548-11e1-b86c-0800200c9a66.ta
+# crypt
+PRODUCT_PACKAGES += cb3e5ba0-adf1-11e0-998b-0002a5d5c51b.ta
+# rpc_test
+PRODUCT_PACKAGES += d17f73a0-36ef-11e1-984a-0002a5d5c51b.ta
+# concurrent
+PRODUCT_PACKAGES += e13010e0-2ae1-11e5-896a-0002a5d5c51b.ta
+# aes_perf
+PRODUCT_PACKAGES += e626662e-c0e2-485c-b8c8-09fbce6edf3d.ta
+# sims
+PRODUCT_PACKAGES += e6a33ed4-562b-463a-bb7e-ff5e15a493c8.ta
+# storage_benchmark
+PRODUCT_PACKAGES += f157cda0-550c-11e5-a6fa-0002a5d5c51b.ta
+# socket
+PRODUCT_PACKAGES += 873bcd08-c2c3-11e6-a937-d0bf9c45c61c.ta
+# sdp-basic
+PRODUCT_PACKAGES += 12345678-5b69-11e4-9dbb-101f74f00099.ta
+
+PRODUCT_PACKAGES += optee_example_hello_world
+PRODUCT_PACKAGES += 8aaaf200-2450-11e4-abe2-0002a5d5c51b.ta
diff --git a/ueventd.common.rc b/ueventd.common.rc
index ff82f70..e5caa05 100644
--- a/ueventd.common.rc
+++ b/ueventd.common.rc
@@ -14,5 +14,8 @@ subsystem usbmisc
 /dev/ion	 0666 system graphics
 /dev/graphics/fb0	0666 system graphics
 
+/dev/tee0        0666   shell     shell
+/dev/teepriv0    0666   shell     shell
+
 /sys/devices/platform/ddr_devfreq/devfreq/ddr_devfreq		min_freq	0644	system	system
 /sys/devices/platform/e82c0000.mali/devfreq/e82c0000.mali	min_freq	0644	system	system
