diff --git a/build_atf.sh b/build_atf.sh
new file mode 100755
index 0000000..f8de719
--- /dev/null
+++ b/build_atf.sh
@@ -0,0 +1,36 @@
+BUILD_OPTION=DEBUG
+
+export ATF_TOOLCHAIN_DIR=$(pwd)/toolchains
+export BUILD_PATH=$(pwd)/..
+export OPTEE_OS_DIR_ATF=${BUILD_PATH}/../optee_os
+export AARCH64_TOOLCHAIN=GCC5
+export UEFI_TOOLS_DIR=${BUILD_PATH}/uefi-tools
+export OPEN_PLATFORM_PKG_DIR=${BUILD_PATH}/OpenPlatformPkg
+export ARM_TRUSTED_FIRMWARE_DIR=${BUILD_PATH}/arm-trusted-firmware
+export EDK2_DIR=${BUILD_PATH}/edk2
+EDK2_OUTPUT_DIR=${EDK2_DIR}/Build/HiKey960/${BUILD_OPTION}_${AARCH64_TOOLCHAIN}
+
+echo $PATH | grep -q ${ATF_TOOLCHAIN_DIR}
+
+if [ "$?" ]; then
+    export PATH=$PATH:${ATF_TOOLCHAIN_DIR}/gcc-linaro-6.4.1-2017.08-x86_64_arm-linux-gnueabihf/bin:${ATF_TOOLCHAIN_DIR}/gcc-linaro-6.4.1-2017.08-x86_64_aarch64-linux-gnu/bin
+fi
+
+cd ${EDK2_DIR}
+ln -sf ${OPEN_PLATFORM_PKG_DIR}
+
+#Build UEFI & ARM Trust Firmware
+${UEFI_TOOLS_DIR}/uefi-build.sh -T ${AARCH64_TOOLCHAIN} -b ${BUILD_OPTION} -a ${ARM_TRUSTED_FIRMWARE_DIR} -s ${OPTEE_OS_DIR_ATF} hikey960
+
+#-  Generate l-loader.bin and partition table.
+#   *Make sure that you're using the sgdisk in the l-loader directory.*
+
+cd ${BUILD_PATH}/l-loader
+ln -sf ${EDK2_OUTPUT_DIR}/FV/bl1.bin
+ln -sf ${EDK2_OUTPUT_DIR}/FV/bl2.bin
+ln -sf ${EDK2_OUTPUT_DIR}/FV/fip.bin
+ln -sf ${EDK2_OUTPUT_DIR}/FV/BL33_AP_UEFI.fd
+
+make hikey960
+
+cd $BUILD_PATH/tools-images-hikey960
